src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlClientEventSource.Windows.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+
+namespace Microsoft.Data.SqlClient
+{
+    internal partial class SqlClientEventSource : EventSource
+    {
+        private bool _traceLoggingProviderEnabled = false;
+
+        /// <summary>
+        /// Captures application flow traces from native networking implementation
+        /// </summary>
+        private const EventCommand SNINativeTrace = (EventCommand)8192;
+
+        /// <summary>
+        /// Captures scope trace events from native networking implementation
+        /// </summary>
+        private const EventCommand SNINativeScope = (EventCommand)16384;
+
+        /// <summary>
+        /// Disables all event tracing in native networking implementation
+        /// </summary>
+        private const EventCommand SNINativeDisable = (EventCommand)32768;
+
+        protected override void OnEventCommand(EventCommandEventArgs e)
+        {
+            // Internally, EventListener.EnableEvents sends an event command, with a reserved value of 0, -2, or -3.
+            // When a command is sent via EnableEvents or SendCommand, check if it is a user-defined value
+            // to enable or disable event tracing in sni.dll.
+            // If registration fails, all write and unregister commands will be a no-op.
+
+            // Only register the provider if it's not already registered. Registering a provider that is already
+            // registered can lead to unpredictable behaviour.
+            if (!_traceLoggingProviderEnabled && e.Command > 0 && (e.Command & (SNINativeTrace | SNINativeScope)) != 0)
+            {
+                int eventKeyword = (int)(e.Command & (SNINativeTrace | SNINativeScope));
+                _traceLoggingProviderEnabled = SNINativeMethodWrapper.RegisterTraceProvider(eventKeyword);
```

Comment:
By default, we use native SNI.dll on Windows. However, we can call the AppContext.SetSwitch() to forcibly use managed SNI on Windows for .net core application. Will this cause some trouble when managed SNI is used on Windows?

Reply:
Good point. If managed networking is enabled, the provider still gets registered but no events are written from SNI.dll. I added another condition in OnEventCommand.